websocket/client.go

```diff
@@ -77,30 +78,48 @@ func parseAuthority(location *url.URL) string {
 	return location.Host
 }
 
-// DialConfig opens a new client connection to a WebSocket with a config.
 func DialConfig(config *Config) (ws *Conn, err error) {
-	var client net.Conn
+	return config.DialWithContext(context.Background())
+}
+
+// DialWithContext opens a new client connection to a WebSocket, with context support for timeouts/cancellation.
+func (config *Config) DialWithContext(ctx context.Context) (ws *Conn, err error) {
 	if config.Location == nil {
 		return nil, &DialError{config, ErrBadWebSocketLocation}
 	}
 	if config.Origin == nil {
 		return nil, &DialError{config, ErrBadWebSocketOrigin}
 	}
+
 	dialer := config.Dialer
 	if dialer == nil {
 		dialer = &net.Dialer{}
 	}
-	client, err = dialWithDialer(dialer, config)
+
+	client, err := dialWithDialer(ctx, dialer, config)
 	if err != nil {
-		goto Error
+		return nil, &DialError{config, err}
 	}
+
+	// The websocket.NewClient() function can block indefinitely, make sure that we
+	// respect the deadlines/cancellations specified by the context.
```

Comment:
https://pkg.go.dev/net#Dialer.DialContext

> The provided Context must be non-nil. If the context expires before the connection is complete, an error is returned. Once successfully connected, any expiration of the context will not affect the connection.

I don't know if the context should apply to the connection itself, it seems it should only affect before the connection